File: Roadmap/02 - FUNCIONES Y ALCANCE/java/willr30.java
public class Solucion_02_Java {

    // Variable global
    static String esto_es_una_variable_global = "Variable_local";

    // Método sin parámetros
    public static void funcion_sin_parametros() {
        System.out.println("Esto es una función sin parámetros");
    }

    // Método que recibe un parámetro
    public static void funcion_con_parametros(String nombre) {
        System.out.println("Esto es un saludo para => " + nombre);
    }

    // Método con un parámetro y tipo de dato definido
    public static void funcion_con_tipo_de_datos_definido(String ciudad) {
        System.out.println("Quiero viajar a " + ciudad);
    }

    // Método con parámetros definidos y tipos de datos
    public static void funcion_con_parametros_definidos_suma(int a, int b) {
        System.out.println("El resultado es " + (a + b));
    }

    // Método con retorno y parámetro definido
    public static String funcion_con_retorno(String comida) {
        return "Quisiera probar " + comida + " algún día";
    }

    // Método que muestra una variable global
    public static void funcion_mostrando_variable_global() {
        System.out.println(esto_es_una_variable_global);
    }

    // Método que muestra las tablas de multiplicar según una variable local
    public static void funcion_con_variable_local() {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.print("Escribe una tabla de multiplicar => ");
        int tabla_de_multiplicar = scanner.nextInt();

        for (int i = 0; i < 12; i++) {
            System.out.println(tabla_de_multiplicar + " * " + i + " = " + (tabla_de_multiplicar * i));
        }
        scanner.close();
    }


    /*
    * Extra
    * */

    public static int extra(String texto_1, String texto_2) {
        int contador = 0;
        // Mostramos los números del 1 al 100
        for (int i = 1; i <= 100; i++) {
            if (i % 3 == 0 && i % 5 == 0) {
                System.out.println(texto_1 + " y " + texto_2);
            } else if (i % 3 == 0) {
                System.out.println(texto_1);
            } else if (i % 5 == 0) {
                System.out.println(texto_2);
            } else {
                System.out.println(i);
                contador++;
            }
        }
        return contador;
    }
    // Método main para probar los demás métodos
    public static void main(String[] args) {
        funcion_sin_parametros();
        funcion_con_parametros("Juan");
        funcion_con_tipo_de_datos_definido("Leon");
        funcion_con_parametros_definidos_suma(3, 5);
        String resultado_retorno = funcion_con_retorno("Nacatamal");
        System.out.println(resultado_retorno);
        funcion_mostrando_variable_global();
        funcion_con_variable_local();

        //llamamos a la funcion extra
        System.out.println(extra("Mi abuelito me ama", "Mi abuelita me ama"));
    }
}

File: Roadmap/02 - FUNCIONES Y ALCANCE/python/willr30.py
'''
/*
 * EJERCICIO:
 * - Crea ejemplos de funciones básicas que representen las diferentes
 *   posibilidades del lenguaje:
 *   Sin parámetros ni retorno, con uno o varios parámetros, con retorno...
 * - Comprueba si puedes crear funciones dentro de funciones.
 * - Utiliza algún ejemplo de funciones ya creadas en el lenguaje.
 * - Pon a prueba el concepto de variable LOCAL y GLOBAL.
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *   (y tener en cuenta que cada lenguaje puede poseer más o menos posibilidades)
 *
 */

'''

#definimos una variable global
esto_es_una_variable_global="Variable_local"

#funcion sin parametros
def funcion_sin_parametros():
    print("Esto es una funcion sin parámetros")

#Funcion que recibe un parámetro
def funcion_con_parametros(nombre):
    print(f"Esto es un saludo para => {nombre}")

#funcion con un parámetro y el tipo de dato definido
def funcion_con_tipo_de_datos_definido(ciudad : str):
    print(f"Quiero viajar a {ciudad}")

#funcion con parametros definidos y tipos de datos
def funcion_con_parametros_definidos_suma(a:int = 2, b: int = 0):
    print(f"El resultaod es {a+b}")

#funcion con retorno y parámetro definido
def funcion_con_retorno(comida:str):
    return f"Quisier probar {comida} algún día"

def funcion_mostrando_variable_global():
    print(f"{esto_es_una_variable_global}")

#funcion que mustra las tablas de multiplcar segun una variable local (que solo existe dentro de la funcion)
def funcion_con_variable_local():
    tabla_de_multiplicar=input("Escribe una tabla de multiplicar => ")

    for i in range (0,12):
        print(f"{tabla_de_multiplicar} * {i} = {tabla_de_multiplicar*i}")


# Llamada a las funciones
funcion_sin_parametros()
funcion_con_parametros("Juan")
funcion_con_tipo_de_datos_definido("Managua")
funcion_con_parametros_definidos_suma(3, 5)

resultado_retorno = funcion_con_retorno("Gallo pinto")
print(resultado_retorno)

funcion_mostrando_variable_global()
funcion_con_variable_local()
#Extra

"""
DIFICULTAD EXTRA (opcional):
 * Crea una función que reciba dos parámetros de tipo cadena de texto y retorne un número.
 * - La función imprime todos los números del 1 al 100. Teniendo en cuenta que:
 *   - Si el número es múltiplo de 3, muestra la cadena de texto del primer parámetro.
 *   - Si el número es múltiplo de 5, muestra la cadena de texto del segundo parámetro.
 *   - Si el número es múltiplo de 3 y de 5, muestra las dos cadenas de texto concatenadas.
 *   - La función retorna el número de veces que se ha impreso el número en lugar de los textos.
"""

def extra(texto_1:str, text_2:str):
    contador=0
    #mostramos los numeros del 1 al 100
    for i in range (0,100):
        if i % 3 :
            print(texto_1)
        elif i % 5:
            print(text_2)
        elif i%3 and i%5:
            print(f"{texto_1} {text_2}")
        else:
            #si las condiciones anteriores no se cumplen se mostrarpa en pantalla el numero
            print(i)
            contador+=1

    return contador



print(extra("primer texto", "segundo texto"))